### PR TITLE
[MPLUGIN-508] Upgrade to Maven 4.0.0-alpha-13

### PR DESCRIPTION
**Pull request by gnodet**
JIRA issue: [MPLUGIN-508](https://issues.apache.org/jira/browse/MPLUGIN-508)

### CODE REVIEW

**Review by cstamas — APPROVED** [review]

File: maven-plugin-plugin/src/it/beanshell-simple/pom.xml [hunk 1]
```diff
@@ -53,6 +53,7 @@
         <artifactId>maven-plugin-plugin</artifactId>
         <version>${pluginPluginVersion}</version>
         <configuration>
+          <requiredMavenVersion>3.6.5</requiredMavenVersion>
```
Comment (slawekjaranowski) [inline maven-plugin-plugin/src/it/beanshell-simple/pom.xml:56]: Maven version 3.6.5 is not exist ... last in 3.6.x is 3.6.3

Comment (gnodet) [reply maven-plugin-plugin/src/it/beanshell-simple/pom.xml]: Fixed

**Comment by hboutemy** [comment]
I don't understand the objective? what does "Upgrade to Maven 4.0.0-alpha-12" mean for this plugin that is for Maven 3?
can you elaborate, please?

**Comment by gnodet** [comment]
> I don't understand the objective? what does "Upgrade to Maven 4.0.0-alpha-12" mean for this plugin that is for Maven 3? can you elaborate, please?

The plugin currently supports building either a Maven 3 or a Maven 4 plugin.  Though I wonder if a streamlined would be better located inside Maven core...

**Comment by hboutemy** [comment]
> The plugin currently supports building either a Maven 3 or a Maven 4 plugin.

sorry, I still need some explanations because I can't guess from this PR touching 75 files described as "Upgrade to Maven 4.0.0-alpha-12"

is this PR about adding support for Maven 4 specific plugins in addition to classical Maven 3 plugins? or just upgrading such a support added before that I overlooked?

Is there a Maven 4 plugin IT to look at such a Maven 4 plugin?

and key question: is this PR intended for 3.12.0 release, as noted in Jira?

**Comment by gnodet** [comment]
> > The plugin currently supports building either a Maven 3 or a Maven 4 plugin.
> 
> sorry, I still need some explanations because I can't guess from this PR touching 75 files described as "Upgrade to Maven 4.0.0-alpha-12"

Most of the 75 files are just a renaming of the `mavenVersion` to `maven3Version` to make it more intuitive which one is used.

> 
> is this PR about adding support for Maven 4 specific plugins in addition to classical Maven 3 plugins? or just upgrading such a support added before that I overlooked?

The support was added quite some time ago (https://github.com/apache/maven-plugin-tools/commit/50482ba4695f2e3efe710a583a639fa53dbabb64).  This PR is just to upgrade to alpha-13 (and needs to be adjusted with the release)

> Is there a Maven 4 plugin IT to look at such a Maven 4 plugin?

Yes, there's one at https://github.com/apache/maven-plugin-tools/tree/5297a91a8f8b4df54c96566b0629bb72944c2047/maven-plugin-plugin/src/it/v4api

> and key question: is this PR intended for 3.12.0 release, as noted in Jira?

Yes

**Comment by gnodet** [comment]
> > > The plugin currently supports building either a Maven 3 or a Maven 4 plugin.
> > 
> > 
> > sorry, I still need some explanations because I can't guess from this PR touching 75 files described as "Upgrade to Maven 4.0.0-alpha-12"
> 
> Most of the 75 files are just a renaming of the `mavenVersion` to `maven3Version` to make it more intuitive which one is used.
> 
> > is this PR about adding support for Maven 4 specific plugins in addition to classical Maven 3 plugins? or just upgrading such a support added before that I overlooked?
> 
> The support was added quite some time ago ([50482ba](https://github.com/apache/maven-plugin-tools/commit/50482ba4695f2e3efe710a583a639fa53dbabb64)). This PR is just to upgrade to alpha-13 (and needs to be adjusted with the release)
> 
> > Is there a Maven 4 plugin IT to look at such a Maven 4 plugin?
> 
> Yes, there's one at https://github.com/apache/maven-plugin-tools/tree/5297a91a8f8b4df54c96566b0629bb72944c2047/maven-plugin-plugin/src/it/v4api
> 
> > and key question: is this PR intended for 3.12.0 release, as noted in Jira?
> 
> Yes

FWIW, I think the plugin supporting Maven 4 would be better located inside maven-core colocated with the API (and maybe soon the implementation extracted with https://github.com/apache/maven/pull/1441).  Same for the plugin-testing framework also contains support for both Maven 3 and Maven 4 in the same library, which is not really clean imho.

**Comment by hboutemy** [comment]
I extracted from this huge PR updates that were not really related to Maven 4 alpha but cleanup that caused it hard to get the intent of the PR
you'll need to rebase to get a more focused PR, that we'll be able to describe: from what I see, the upgrade adds some very new approach like DI that would be useful to describe

I'd like to do a release next week: we'll need to define if we can get this update in or not

**Comment by gnodet** [comment]
> I extracted from this huge PR updates that were not really related to Maven 4 alpha but cleanup that caused it hard to get the intent of the PR you'll need to rebase to get a more focused PR, that we'll be able to describe: from what I see, the upgrade adds some very new approach like DI that would be useful to describe
> 
> I'd like to do a release next week: we'll need to define if we can get this update in or not

The support for 4.x plugins cannot really be finalised before the API is, given this target is moving.  However, we need to keep up to date so that we can work on plugins more easily.  I think the plugin could be trimmed down (just the actual plugin, we don't need other modules) and moved to maven-core to support v4 plugins...

**Comment by hboutemy** [comment]
too complex answer, I'm not sure what to do to release plugin-tools 3.12.0 this WE regarding this PR associated to MPLUGIN-508:
1. release now without the PR?
2. push to finalize to PR before releasing? (need to define the acceptance criteria)

**Comment by hboutemy** [comment]
IIUC, discussion means this PR is not ready for merge and release: I'll release 3.12.0 without

File: pom.xml [hunk 2]
```diff
@@ -333,6 +333,21 @@
...
+      <releases>
+        <enabled>true</enabled>
+        <updatePolicy>never</updatePolicy>
+        <checksumPolicy>fail</checksumPolicy>
+      </releases>
+      <snapshots>
+        <enabled>false</enabled>
+      </snapshots>
+      <id>maven-2126</id>
+      <url>https://repository.apache.org/content/repositories/maven-2126/</url>
```
Comment (slawekjaranowski) [inline pom.xml:347]: should be a PR as draft

**Comment by gnodet** [comment]
Closed in favour of https://github.com/apache/maven-plugin-tools/pull/286

**Comment by jira-importer** [comment]
Resolve #826

**Pull request closed without merging**